README.md

```diff
@@ -455,15 +455,26 @@ The list of configurations for various `fms_acceleration` plugins:
 - [fused_ops_and_kernels](./tuning/config/acceleration_configs/fused_ops_and_kernels.py) (experimental):
   - `--fused_lora`: fused lora for more efficient LoRA training.
   - `--fast_kernels`: fast cross-entropy, rope, rms loss kernels.
+- [attention_and_distributed_packing](./tuning/config/acceleration_configs/attention_and_distributed_packing.py) (experimental):
+  - `--padding_free`: technique to process multiple examples in single batch without adding padding tokens that waste compute.
+  - `--multipack`: technique for *multi-gpu training* to balance out number of tokens processed in each device, to minimize waiting time.
 
 Notes: 
  * `quantized_lora_config` requires that it be used along with LoRA tuning technique. See [LoRA tuning section](https://github.com/foundation-model-stack/fms-hf-tuning/tree/main?tab=readme-ov-file#lora-tuning-example) on the LoRA parameters to pass.
  * When setting `--auto_gptq triton_v2` plus note to also pass `--torch_dtype float16` and `--fp16`, or an exception will be raised. This is because these kernels only support this dtype.
- * Currently, the `fused_ops_and_kernels` is to be used used together QLoRA or GPTQ-LORA via the `quantized_lora_config`. In the future it may be made more flexible such that `fast_kernels` can even be used with full-finetuning.
  * When using `fused_ops_and_kernels` together with `quantized_lora_config`,
  make sure to appropriately set `--fused_lora auto_gptq True` or `bitsandbytes True`; the `True` sets `fast_lora==True`.
- * Currently `fused_ops_and_kernels` only supports activating `fast_loss,fast_rsm_layernorm,fast_rope_embeddings` all to `True`, so pass `--fast_kernels True True True`.
-
+ * `fused_ops_and_kernels` works for full-finetuning, LoRA, QLoRA and GPTQ-LORA, 
+    - pass `--fast_kernels True True True` for full finetuning/LoRA
+    - pass `--fast_kernels True True True --auto_gptq triton_v2 --fused_lora auto_gptq True` for GPTQ-LoRA
+    - pass `--fast_kernels True True True --bitsandbytes nf4 --fused_lora bitsandbytes True` for QLoRA
```

Comment:
Yes that is correct, but unfortunately that will be more complicated than the current implementation. 
- Consider the plugin `dataclass` (e.g., `FusedOpsAndKernelsConfig`), see [here](https://github.com/foundation-model-stack/fms-hf-tuning/blob/main/tuning/config/acceleration_configs/fused_ops_and_kernels.py#L72) 
- the plugin dataclass is a nested `dataclass`; this is because it has `dataclasses` as members.
- each member dataclass needs to be parsable by `HfArgumentParser`, which actually does not support parsing a `dataclass` type.
- hence, we made it possible due to our  [parsable_dataclass](https://github.com/foundation-model-stack/fms-hf-tuning/blob/main/tuning/config/acceleration_configs/utils.py#L62) decorator, that
  - masquarades the member dataclass as a `List`, which `HfArgumentParser` does support lists of a *uniform type*.
  - allows our member dataclass to contain *mixed types* by the casting logic implemented in `parsable_dataclass` via [EnsureTypes](https://github.com/foundation-model-stack/fms-hf-tuning/blob/main/tuning/config/acceleration_configs/utils.py#L38C7-L38C18).

Hence this allows `--fast_kernels False True False` to be parsed into the dataclass `FusedOpsAndKernelsConfig`. 

To support parsing of the kind `--fast_kernels FastRoPE`, we need to handle
- handling of different types. "FastRoPE" is clearly a boolean type, but we also need to handle `str` inputs, `float` inputs etc, where it would need to be a `key=value` pair
- handling different orders, we need to be able to parse `--dataclass_key key_a=a key_b` and `--dataclass key_b key_a=a` equivalently.
- handling defaults, if `key_a` is missing then we need to set it a default. We need to raise if users try to define the dataclass without defaults.

Reply:
I suggest to merge this PR first, we can come back to this again when we do the fused cross entropy. BTW i left a comment for @achew010 to help upgrade `FusedOpsAndKernelsConfig` to non-experimental status in this PR by deleting these `experimental=True` entries, see [here](https://github.com/foundation-model-stack/fms-hf-tuning/blob/main/tuning/config/acceleration_configs/acceleration_framework_config.py#L86)

Reply:
Thank you for the details, I agree getting this merged and thinking about improvements for fast_kernels later makes sense. Is just the `FusedOpsAndKernelsConfig` ready to move out of experimental or can this also be done for PaddingFree and Multipack?